sqlx-postgres/src/copy.rs

```diff
@@ -358,3 +362,192 @@ async fn pg_begin_copy_out<'c, C: DerefMut<Target = PgConnection> + Send + 'c>(
 
     Ok(Box::pin(stream))
 }
+
+enum PgCopyBothCommand {
+    Begin(Vec<u8>),
+    CopyData(Vec<u8>),
+    CopyDone { from_client: bool },
+}
+
+pub struct PgCopyBothSender(flume::Sender<PgCopyBothCommand>);
+pub struct PgCopyBothReceiver(flume::r#async::RecvStream<'static, Result<Bytes>>);
+
+// Open a duplex connection allowing high-speed bulk data transfer to and from the server.
+//
+// # Example
+//
+// ```rust,no_run
+// # use sqlx_postgres::{
+// #     pg_copy_both, PgConnectOptions, PgPoolOptions, PgReplicationMode,
+// # };
+// // Connection must be configured with a replication mode!
+// let options = PgConnectOptions::new()
+//     .host("0.0.0.0")
+//     .replication_mode(PgReplicationMode::Logical);
+//
+// let pool = PgPoolOptions::new()
+//     .connect_with(options)
+//     .await
+//     .expect("failed to connect to postgres");
+//
+// let query = format!(
+//     r#"START_REPLICATION SLOT "{}" LOGICAL {} ("proto_version" '1', "publication_names" '{}')"#,
+//     "test_slot", "0/1573178", "test_publication",
+// );
+// let (copy_tx, copy_recv) = pg_copy_both(&pool, query.as_str())
+//     .await
+//     .expect("start replication");
+// // Read data from the server.
+// while let Some(data) = copy_recv.next().await {
+//     println!("data: {:?}", data);
+//     // And send some back (e.g. keepalive).
+//     copy_tx.send(Vec::new()).await?;
+// }
+// // Connection closed.
+// ```
+pub async fn pg_copy_both(
+    pool: &Pool<Postgres>,
+    statement: &str,
+) -> Result<(PgCopyBothSender, PgCopyBothReceiver), Error> {
+    // Setup upstream/downstream channels.
+    let (recv_tx, recv_rx) = flume::bounded(1);
+    let (send_tx, send_rx) = flume::bounded(1);
+
+    crate::rt::spawn({
+        let pool = pool.clone();
+        async move {
+            if let Err(err) = copy_both_handler(pool, recv_tx.clone(), send_rx).await {
+                let _ignored = recv_tx.send_async(Err(err)).await;
+            }
+        }
+    });
+
+    // Execute the given statement to switch into CopyBoth mode.
+    let mut buf = Vec::new();
+    Query(statement).encode(&mut buf);
+    send_tx
+        .send_async(PgCopyBothCommand::Begin(buf))
+        .await
+        .map_err(|_err| Error::WorkerCrashed)?;
+
+    Ok((
+        PgCopyBothSender(send_tx),
+        PgCopyBothReceiver(recv_rx.into_stream()),
+    ))
+}
+
+impl PgCopyBothSender {
+    /// Send a chunk of `COPY` data.
+    pub async fn send(&self, data: impl Into<Vec<u8>>) -> Result<()> {
+        self.0
+            .send_async(PgCopyBothCommand::CopyData(data.into()))
```

Comment:
Another reason not to do the sender/receiver split at this level, `send_async()` is not strictly cancel-safe: https://github.com/zesterer/flume/issues/104#issuecomment-1218515430

Reply:
Hum good point!

How can I avoid/solve this?

Reply:
Any suggestion/pointer @abonander ?